plasmapy/simulation/particletracker.py

```diff
@@ -1,203 +1,134 @@
-
-    def boris_push(self, init=False):
-        r"""
-        Implements the Boris algorithm for moving particles and updating their
-        velocities.
-
-        Arguments
-        ----------
-        init : bool (optional)
-            If `True`, does not change the particle positions and sets dt
-            to -dt/2.
-
-        Notes
-        ----------
-        The Boris algorithm is the standard energy conserving algorithm for
-        particle movement in plasma physics. See [1]_ for more details.
-
-        Conceptually, the algorithm has three phases:
+    The Boris algorithm is the standard energy conserving algorithm for
+    particle movement in plasma physics. See [1]_ for more details.
 
-        1. Add half the impulse from electric field.
-        2. Rotate the particle velocity about the direction of the magnetic
-           field.
-        3. Add the second half of the impulse from the electric field.
+    Conceptually, the algorithm has three phases:
 
-        This ends up causing the magnetic field action to be properly
-        "centered" in time, and the algorithm conserves energy.
+    1. Add half the impulse from electric field.
+    2. Rotate the particle velocity about the direction of the magnetic
+       field.
+    3. Add the second half of the impulse from the electric field.
 
-        References
-        ----------
-        .. [1] C. K. Birdsall, A. B. Langdon, "Plasma Physics via Computer
-               Simulation", 2004, p. 58-63
-        """
-        dt = -self.dt / 2 if init else self.dt
-        b, e = self._interpolate_fields()
+    This ends up causing the magnetic field action to be properly
+    "centered" in time, and the algorithm conserves energy.
 
+    References
+    ----------
+    .. [1] C. K. Birdsall, A. B. Langdon, "Plasma Physics via Computer
+           Simulation", 2004, p. 58-63
+    """
+    for i in numba.prange(len(x)):
         # add first half of electric impulse
-        vminus = self.v + self.eff_q * e / self.eff_m * dt * 0.5
+        vminus = v[i] + hqmdt * e[i]
 
         # rotate to add magnetic field
-        t = -b * self.eff_q / self.eff_m * dt * 0.5
-        s = 2 * t / (1 + (t * t).sum(axis=1, keepdims=True))
-        vprime = vminus + np.cross(vminus.si.value, t) * u.m / u.s
-        vplus = vminus + np.cross(vprime.si.value, s) * u.m / u.s
+        t = -b[i] * hqmdt
+        s = 2 * t / (1 + (t[0] * t[0] + t[1] * t[1] + t[2] * t[2]))
+        cross_result = np.cross(vminus, t)
+        vprime = vminus + cross_result
+        cross_result_2 = np.cross(vprime, s)
+        vplus = vminus + cross_result_2
 
         # add second half of electric impulse
-        v_new = vplus + self.eff_q * e / self.eff_m * dt * 0.5
+        v[i] = vplus + e[i] * hqmdt
+        x[i] += v[i] * dt
 
-        self.v = v_new
-        if not init:
-            self.x += self.v * dt
+class ParticleTrackerSolution:
+    """
+    A solution to `ParticleTracker`'s trajectory integration.
```

Comment:
I like having a separate class for the particle tracker solution.

Reply:
Picked it up from scipy solve ivp and pymc3 outputs. Pymc3 is actually nice in that Hamiltonian Monte Carlo is actually a particle simulation with wacky momenta in model parameter space, so there's plenty tying it back to usual particle stuff and nonlinear dynamics. It's absolutely lovely.